### PR TITLE
ssvsigner: break circular dependency with ssv-node

### PR DESCRIPTION
**Pull request by nkryuchkov**
as per title

Closes ssvlabs/ssv-node-board#1057

### CODE REVIEW

**Comment by nkryuchkov** [comment]
@greptileai please review the PR

**Comment by nkryuchkov** [comment]
@greptileai review this again, considering all comments on this PR

**Comment by nkryuchkov** [comment]
@greptileai review this again, considering all comments on this PR

File: ekmadapter/database.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+func (a *databaseAdapter) GetAll(txn ekm.ReadTxn, prefix []byte, handler func(int, ekm.Obj) error) error {
+	reader, err := asReader(txn)
+	if err != nil {
+		return err
+	}
+
+	return a.db.UsingReader(reader).GetAll(prefix, func(i int, obj basedb.Obj) error {
+		return handler(i, ekm.Obj{Key: obj.Key, Value: obj.Value})
+	})
+}
```
Comment (iurii-ssv) [inline ekmadapter/database.go:44]: Probably better to keep these method-definition next to the `databaseAdapter` definition (so it's easy to find all struct-related methods)

File: ekmadapter/database.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	"fmt"
+
+	"github.com/ssvlabs/ssv/ssvsigner/ekm"
+	"github.com/ssvlabs/ssv/storage/basedb"
+)
+
+// NewDatabaseAdapter bridges node's basedb.Database to ekm.Database.
+func NewDatabaseAdapter(db basedb.Database) ekm.Database {
+	return &databaseAdapter{db: db}
+}
```
Comment (iurii-ssv) [inline ekmadapter/database.go:13]: I'm not sure I understand why this package is:
- not residing under `ssvsigner` directory (it does reference things like `ekm.Database` that are under `ssvsigner` ... so might as well move it there too)
- has "adapter" in the package/primitives names ... I see that it basically "converts one interface to another" but seems like we could just call it/use it as `ekm.NewDatabase(...` passing in the underlying storage too (the "adapter" naming adds unnecessary complexity IMO)

**Review by iurii-ssv — DISMISSED** [review]
Good stuff!

File: ekmadapter/database.go (hunk 2 above)
Comment (nkryuchkov) [reply ekmadapter/database.go:13]: This package stays in `ssv` intentionally because it depends on node-only `storage/basedb`, while `ssvsigner` must remain importable without any dependency back on the root module.

The adapter naming is also intentional: its job is specifically to translate the node's `basedb.Database` into `ekm.Database` without teaching `ekm` about root-module storage types. Moving this into `ssvsigner` or folding it into `ekm` would reintroduce the boundary leak we split out in this PR.

Comment (nkryuchkov) [reply ekmadapter/database.go:13]: Added a `doc.go` with an explanation

Comment (iurii-ssv) [reply ekmadapter/database.go:13]: Got it.

File: Makefile [hunk 3]
```diff
@@ -53,11 +53,13 @@ full-test:
 	@echo "Running all tests"
 	@go test -tags blst_enabled -timeout 20m ${COV_CMD} -p 1 ./...
 	@cd ssvsigner && go test -tags blst_enabled -timeout 20m ${COV_CMD} -p 1 ./...
-
 .PHONY: unit-test
```
Comment (iurii-ssv) [inline Makefile]: Seems like we should keep this newline (typo, I assume) ?

**Review by iurii-ssv — DISMISSED** [review]
Just 1 nit

**Review by olegshmuelov — DISMISSED** [review]
LGTM.

One nit: the new `full-test` target doesn't include `-race`, `unit-test` does. Worth adding for consistency.

**Comment by nkryuchkov** [comment]
> One nit: the new `full-test` target doesn't include `-race`, `unit-test` does. Worth adding for consistency.

I accidentally added `full-test` while merging, although we don't need it. Deleted

File: Makefile (hunk 3 above)
Comment (nkryuchkov) [reply Makefile]: We need to delete `full-test`, I fixed a merge conflict here incorrectly

**Review by iurii-ssv — DISMISSED** [review]

**Review by olegshmuelov — DISMISSED** [review]

**Review by iurii-ssv — DISMISSED** [review]

File: scripts/ssvsigner_boundary.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+set -eo pipefail
+
+echo -e "\033[36mINFO\033[0m [ssvsigner-boundary] Verifying ssvsigner does not import root ssv packages"
+
+OUT="$(
+  cd ssvsigner
+  go list -f '{{range .Imports}}{{println .}}{{end}}{{range .TestImports}}{{println .}}{{end}}{{range .XTestImports}}{{println .}}{{end}}' ./... \
+    | sort -u \
+    | grep '^github.com/ssvlabs/ssv/' \
+    | grep -Ev '^github.com/ssvlabs/ssv/ssvsigner($|/)' || true
```
Comment (olegshmuelov) [inline scripts/ssvsigner_boundary.sh]: The `|| true` at the end of the pipeline prevents failure when `grep` finds no forbidden imports (the success case), but it also swallows `go list` failures. If `go list` errors out (syntax error, unresolved dependency), `OUT` is empty and the script reports "Import boundary check passed" — making the guard fail-open.

Low severity because CI also runs build and tests which would catch broken packages, but the boundary check itself is unreliable in that scenario.

Suggestion:
```bash
cd ssvsigner

# Fail fast if go list itself errors.
ALL_IMPORTS="$(go list -f '{{range .Imports}}{{println .}}{{end}}{{range .TestImports}}{{println .}}{{end}}{{range .XTestImports}}{{println .}}{{end}}' ./...)"

OUT="$(echo "$ALL_IMPORTS" \
  | sort -u \
  | grep '^github.com/ssvlabs/ssv/' \
  | grep -Ev '^github.com/ssvlabs/ssv/ssvsigner($|/)' || true
)"
```

**Review by olegshmuelov — DISMISSED** [review]
One minor comment

File: ssvsigner/ekm/contracts.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+// NetworkContext groups network parameters used by key managers.
+// Name must match the network identifier represented by Beacon (for example
+// networkconfig.Beacon.Name), because Name is used as the storage key prefix
+// for slashing-protection data.
+type NetworkContext struct {
+	Name                  string
+	Beacon                BeaconNetwork
+	GenesisValidatorsRoot phase0.Root
+}
```
Comment (ljuba-ssv) [inline ssvsigner/ekm/contracts.go]: Nit: before this PR, a Name/Beacon mismatch was structurally impossible (single `*networkconfig.Beacon` carried both). Now they are separate fields set by hand at ~8 call sites. A mismatch would silently write slashing-protection data under the wrong storage prefix.

All current call sites are correct (`Name` always derived from the same config as `Beacon`), so no active bug. But since the consequence is severe (bypassed slashing protection), consider a constructor that validates the coupling, or at minimum, make `Name` a computed property if `BeaconNetwork` can expose it.

**Review by ljuba-ssv — DISMISSED** [review]

**Review by iurii-ssv — APPROVED** [review]

**Review by ljuba-ssv — APPROVED** [review]
